pkg/operator/podsecurityreadinesscontroller/violation.go

```diff
@@ -0,0 +1,130 @@
+package podsecurityreadinesscontroller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/sets"
+	applyconfiguration "k8s.io/client-go/applyconfigurations/core/v1"
+	"k8s.io/klog/v2"
+	psapi "k8s.io/pod-security-admission/api"
+)
+
+const (
+	syncerControllerName = "pod-security-admission-label-synchronization-controller"
+)
+
+var (
+	alertLabels = sets.Set[string]{
+		psapi.WarnLevelLabel:  {},
+		psapi.AuditLevelLabel: {},
+	}
+)
+
+func (c *PodSecurityReadinessController) isNamespaceViolating(ctx context.Context, ns *corev1.Namespace) (bool, error) {
+	viableLabels := getSyncerLabels(ns.ManagedFields).Intersection(alertLabels)
+	if viableLabels.Len() == 0 {
+		// If there are no labels managed by the syncer, we can't make a decision.
+		return false, nil
+	}
+
+	nsApply := applyconfiguration.Namespace(ns.Name).WithLabels(map[string]string{
+		psapi.EnforceLevelLabel: pickStrictest(ns, viableLabels),
+	})
+
+	_, err := c.kubeClient.CoreV1().
+		Namespaces().
+		Apply(ctx, nsApply, metav1.ApplyOptions{
+			DryRun:       []string{metav1.DryRunAll},
+			FieldManager: "pod-security-readiness-controller",
+		})
+	if err != nil {
+		return false, err
+	}
+
+	// If there are warnings, the namespace is violating.
+	return len(c.warningsHandler.PopAll()) > 0, nil
+}
+
+func pickStrictest(ns *corev1.Namespace, viableLabels sets.Set[string]) string {
+	targetLevel := ""
+	for label := range viableLabels {
+		levelStr, ok := ns.Labels[label]
+		if !ok {
+			continue
+		}
+
+		level, err := psapi.ParseLevel(levelStr)
+		if err != nil {
+			klog.V(4).InfoS("invalid level", "namespace", ns.Name, "level", levelStr)
+			continue
+		}
+
+		if targetLevel == "" {
+			targetLevel = levelStr
+			continue
+		}
+
+		if psapi.CompareLevels(psapi.Level(targetLevel), level) < 0 {
+			targetLevel = levelStr
+		}
+	}
+
+	if targetLevel == "" {
+		// Global Config will set it to "restricted", but shouldn't happen.
+		return string(psapi.LevelRestricted)
+	}
+
+	return targetLevel
+}
+
+func getSyncerLabels(managedFields []metav1.ManagedFieldsEntry) sets.Set[string] {
+	ownedLabels := sets.New[string]()
+
+	for _, entry := range managedFields {
+		if entry.Manager != "" && entry.Manager != syncerControllerName {
```

Comment:
It might be that we only care about `syncerControllerName` cases as this is the controller setting it... 😅